mimic/model/identity.py

```diff
@@ -20,3 +26,205 @@ def from_store(cls, session_store, tenant_id):
+    """
+    An object representing username/password + optional tenant credentials
+    """
+    username = attr(validator=validators.instance_of(string_types))
+    password = attr(validator=validators.instance_of(string_types))
+    tenant_id = attr(default=None)
+
+    def get_session(self, session_store):
+        """
+        Get a session corresponding to the user and tenant from the given
+        session store.
+
+        :param session_store: a :class:`mimic.session.SessionStore`
+        """
+        return session_store.session_for_username_password(
+            self.username, self.password, self.tenant_id)
+
+    @classmethod
+    def from_json(cls, json_blob):
+        """
+        Given an authentication JSON blob, which should look like:
+
+            {
+                "auth": {
+                    "passwordCredentials": {
+                        "username": "user",
+                        "password": "pass"
+                    },
+                    "tenantId": "111111"
+                }
+            }
+
+        ``"tenantId"`` is interchanable with ``"tenantName"``, and is
+        optional.
+
+        :return: a class:`PasswordCredentials` object
+        """
+        tenant_id = (json_blob['auth'].get('tenantName', None) or
+                     json_blob['auth'].get('tenantId', None))
+        username = json_blob['auth']['passwordCredentials']['username']
+        password = json_blob['auth']['passwordCredentials']['password']
+        return cls(username=username, password=password, tenant_id=tenant_id)
+
+
+@implementer(ICredentials)
+@attributes
+class APIKeyCredentials(object):
+    """
+    An object representing username/api-key + optional tenant credentials
+    """
+    username = attr(validator=validators.instance_of(string_types))
+    api_key = attr(validator=validators.instance_of(string_types))
+    tenant_id = attr(default=None)
+
+    def get_session(self, session_store):
+        """
+        Get a session corresponding to the user and tenant from the given
+        session store.
+
+        :param session_store: a :class:`mimic.session.SessionStore`
+        """
+        return session_store.session_for_api_key(
+            self.username, self.api_key, self.tenant_id)
+
+    @classmethod
+    def from_json(cls, json_blob):
+        """
+        Given an authentication JSON blob, which should look like:
+
+            {
+                "auth": {
+                    "RAX-KSKEY:apiKeyCredentials": {
+                        "username": "user",
+                        "apiKey": "key"
+                    },
+                    "tenantId": "111111"
+                }
+            }
+
+        ``"tenantId"`` is interchanable with ``"tenantName"``, and is
+        optional.
+
+        :return: a class:`APIKeyCredentials` object
+        """
+        creds = json_blob['auth']
+        tenant_id = (creds.get('tenantName', None) or
+                     creds.get('tenantId', None))
```

Comment:
Minor nitpick: if you have a falsey `tenantName` or `tenantId` (`""` or `0` or `None` in JSON-land) this will give effectively a wrong result.  Given that bad results are Mimic's whole bailiwick, it might actually be important to preserve this and have a test for it and stuff.  Possibly in a separate issue.  But for now, you can shorten the implementation by doing `creds.get('tenantName', creds.get('tenantId'))` to give this the proper priority.